### ISSUE TITLE
PE extension methods should not be filtered based on signature

### ISSUE THREAD

**Issue by OmarTawfik** [dotnet/roslyn#20282]
After discussion with @AlekseyTs, there is a private helper `PEMethodSymbol.IsValidExtensionMethodSignature()` that is used to filter out extension methods (with a valid extension attribute) but don't have a valid signature (like having no parameters, or a `params` first receiver parameter).

This method should be deleted, as for purposes of loading symbols, methods with an extension attribute should be considered extension methods. And later on after overload resolution, such methods that don't conform to C# rules on extension methods can be dropped.
This lends to a consistent surface across method symbols that come from the same compilation, different compilations, and binary references.

This behavior manifests in the following bug:

```csharp
// Reference Project
public static class Ext
{
    public static void M(this params int[] x) { }
}

// User Project
class Program
{
    static void Main(string[] args)
    {
        int x = 5;
        x.M();
    }
}
```

The reference project will have an error on the `params` modifier (as expected). But when referencing this:
1) If it was a compilation reference, the user project will have no errors, and the symbol will be used correctly.
2) If the reference was IL, the user project will drop the symbol and will produce a method not found error.

**Comment by AlekseyTs** [comment]
>And later on after overload resolution, such methods that don't conform to C# rules on extension methods can be dropped.

It would probably be better to filter the methods out during lookup.

**Comment by OmarTawfik** [comment]
@AlekseyTs I now believe the issue is bigger. We're allowing symbols with diagnostics in general to be used. For example, the below test passes. We may have to fix this in on a bigger scope:

```csharp

        [Fact]
        public void Test()
        {
            var referenceCode = @"
public static class Extensions
{
    public static void PrintValue(this int p, this int q)
    {
    }
}";

            var reference = CreateCompilationWithMscorlibAndSystemCore(referenceCode).VerifyDiagnostics(
                // (4,47): error CS1100: Method 'PrintValue' has a parameter modifier 'this' which is not on the first parameter
                //     public static void PrintValue(this int p, this int q)
                Diagnostic(ErrorCode.ERR_BadThisParam, "this").WithArguments("PrintValue").WithLocation(4, 47));

           var user = CreateStandardCompilation(@"
public static class Program
{
    public static void Main()
    {
        2.PrintValue(5);
    }
}", references: new[] { reference.ToMetadataReference() }).VerifyDiagnostics().VerifyEmitDiagnostics();

            var assembly = user.EmitToImageReference();
            Assert.NotNull(assembly);
        }
```

**Comment by OmarTawfik** [comment]
The `IsExtensionMethod` API is already used during look up as you suggested:
https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Portable/Symbols/NamedTypeSymbol.cs#L374

```csharp
        internal void DoGetExtensionMethods(ArrayBuilder<MethodSymbol> methods, string nameOpt, int arity, LookupOptions options)
        {
            var members = nameOpt == null
                ? this.GetMembersUnordered()
                : this.GetSimpleNonTypeMembers(nameOpt);

            foreach (var member in members)
            {
                if (member.Kind == SymbolKind.Method)
                {
                    var method = (MethodSymbol)member;
                    if (method.IsExtensionMethod &&
                        ((options & LookupOptions.AllMethodsOnArityZero) != 0 || arity == method.Arity))
                    {
                        Debug.Assert(method.MethodKind != MethodKind.ReducedExtension);
                        methods.Add(method);
                    }
                }
            }
        }
```

**Comment by AlekseyTs** [comment]
> The IsExtensionMethod API is already used during look up as you suggested

I am not sure what suggestion you are referring to, I do not believe I suggested anything about using IsExtensionMethod API in name lookup in relation to this issue.

**Comment by AlekseyTs** [comment]
> We're allowing symbols with diagnostics in general to be used. 

This is correct, but this is not the problem that originated the issue. The problem was about being able to successfully consume APIs from source that wouldn't be consumed from metadata, or vice-versa. It is very likely that the example doesn't fall into the same category because **this** modifier is not part of metadata, it simply isn't there.

**Comment by AlekseyTs** [comment]
To expand. In the first repro, that is used to open this issue, an invocation ```x.M();``` compiles without an error, but it shouldn't because this violates the rules for consuming extension methods. The method extends an array type, not an integer type. The error at the declaration site helps people to avoid declaring methods that couldn't be consumed, but the primary goal is to prevent the erroneous consumption, which error at the declaration doesn't do. Therefore, there must be an error at consumption site, which happened to be reported only when the method is declared in metadata.
